File: ios/Packages/AppFeatures/Sources/HadithFeature/HadithModels.swift
import Foundation

/// Local reading progress for one collection (docs/features/hadith-collections.md).
/// `readNumbers` is a set — re-reading an entry never double counts.
public struct HadithProgress: Codable, Equatable, Sendable {
    public var readNumbers: Set<Int>
    public var lastReadNumber: Int?

    public init(readNumbers: Set<Int> = [], lastReadNumber: Int? = nil) {
        self.readNumbers = readNumbers
        self.lastReadNumber = lastReadNumber
    }
}

public protocol HadithStoring: Sendable {
    func loadProgress() -> [String: HadithProgress]
    func saveProgress(_ progress: [String: HadithProgress])
}

public struct FileHadithStore: HadithStoring {
    private let fileURL: URL

    public init(directory: URL) {
        fileURL = directory.appendingPathComponent("hadith-progress.json")
    }

    public func loadProgress() -> [String: HadithProgress] {
        guard let data = try? Data(contentsOf: fileURL) else { return [:] }
        return (try? JSONDecoder().decode([String: HadithProgress].self, from: data)) ?? [:]
    }

    public func saveProgress(_ progress: [String: HadithProgress]) {
        guard let data = try? JSONEncoder().encode(progress) else { return }
        try? FileManager.default.createDirectory(
            at: fileURL.deletingLastPathComponent(), withIntermediateDirectories: true
        )
        try? data.write(to: fileURL, options: .atomic)
    }
}

File: android/feature/hadith/src/main/kotlin/com/fatwabot/feature/hadith/HadithModels.kt
package com.fatwabot.feature.hadith

import java.io.File
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

/** Local reading progress for one collection — mirror of iOS HadithProgress
 * (docs/features/hadith-collections.md). `readNumbers` is a set — re-reading
 * an entry never double counts. */
@Serializable
data class HadithProgress(
    val readNumbers: Set<Int> = emptySet(),
    val lastReadNumber: Int? = null,
)

interface HadithStoring {
    fun loadProgress(): Map<String, HadithProgress>
    fun saveProgress(progress: Map<String, HadithProgress>)
}

class FileHadithStore(private val file: File) : HadithStoring {
    private val json = Json { ignoreUnknownKeys = true }
    private val serializer = kotlinx.serialization.serializer<Map<String, HadithProgress>>()

    override fun loadProgress(): Map<String, HadithProgress> = runCatching {
        if (!file.exists()) emptyMap() else json.decodeFromString(serializer, file.readText())
    }.getOrDefault(emptyMap())

    override fun saveProgress(progress: Map<String, HadithProgress>) {
        runCatching {
            file.parentFile?.mkdirs()
            val tmp = File(file.parentFile, "${file.name}.tmp")
            tmp.writeText(json.encodeToString(serializer, progress))
            tmp.renameTo(file)
        }
    }
}
